app/Services/EDocument/Imports/ZugferdEDocument.php

```diff
@@ -0,0 +1,116 @@
+use Exception;
+use horstoeko\zugferd\ZugferdDocumentReader;
+use horstoeko\zugferdvisualizer\ZugferdVisualizer;
+
+class ZugferdEDocument extends AbstractService
+{
+    public ZugferdDocumentReader|string $document;
+
+    /**
+     * @throws Exception
+     */
+    public function __construct(public string $tempdocument, public string $documentname)
+    {
+        # curl -X POST http://localhost:8000/api/v1/edocument/upload -H "Content-Type: multipart/form-data" -H "X-API-TOKEN: 7tdDdkz987H3AYIWhNGXy8jTjJIoDhkAclCDLE26cTCj1KYX7EBHC66VEitJwWhn" -H "X-Requested-With: XMLHttpRequest" -F _method=PUT -F documents[]=@einvoice.xml
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function run(): string
+    {
+        $user = auth()->user();
+        nlog("!---------------------!");
+        $this->document = ZugferdDocumentReader::readAndGuessFromContent($this->tempdocument);
+        $this->document->getDocumentInformation($documentno, $documenttypecode, $documentdate, $invoiceCurrency, $taxCurrency, $documentname, $documentlanguage, $effectiveSpecifiedPeriod);
+        nlog($documentno);
+        nlog($documenttypecode);
+        nlog($documentdate);
+        nlog($invoiceCurrency);
+        nlog($taxCurrency);
+        nlog($documentname);
+        nlog($documentlanguage);
+        nlog($effectiveSpecifiedPeriod);
+        $this->document->getDocumentSummation($grandTotalAmount, $duePayableAmount, $lineTotalAmount, $chargeTotalAmount, $allowanceTotalAmount, $taxBasisTotalAmount, $taxTotalAmount, $roundingAmount, $totalPrepaidAmount);
+        nlog($grandTotalAmount);
+        nlog($duePayableAmount);
+
+        $expenses = Expense::all();
+        // Check if the document already exists as an expense
+        $existingExpense = $expenses->first(function ($expense) use ($documentno, $grandTotalAmount, $documentdate) {
+            return $expense->transaction_reference == $documentno && $expense->amount == $grandTotalAmount && $expense->date == $documentdate;
+        });
+
+        if ($existingExpense) {
+            // The document already exists as an expense
+            return $existingExpense;
+        } else {
+            // The document does not exist as an expense
+            // Handle accordingly
+            $visualizer = new ZugferdVisualizer($this->document);
+            $visualizer->setDefaultTemplate();
+            $visualizer->setPdfFontDefault("arial");
+            $visualizer->setPdfPaperSize('A4-P');
+
+            $expense = ExpenseFactory::create($user->company()->id, $user->id);
+            $expense->date = $documentdate;
+            $expense->user_id = $user->id;
+            $expense->company_id = $user->company()->id;
+            $expense->public_notes = $documentno;
+            $expense->currency_id = Currency::whereCode($invoiceCurrency)->first()->id;
+            $expense->save();
+            $expense->documents()->create(["content" => $visualizer->renderPdf()]);
```

Comment:
@turbo124 I'm having my difficulties with this line. How can I save a string-file to the documents of an expense?

Reply:
i think you should just use the UploadedFile object from the request instead of a stringfile.

i also added some helperclasses with my PR at:
app/Utils/TempFile.php